### ISSUE TITLE
Regression in 2.0: `cabal v1-install happy` fails with mysterious interface error

### ISSUE THREAD

**Issue by andreasabel** [haskell/happy#315]
Lifted from:
- https://github.com/agda/agda/issues/7507

CI failure due to happy-2.0: https://github.com/agda/agda-stdlib/actions/runs/11030950764/job/30636656269#step:6:34
```
$ cabal v1-install happy
...
Building executable 'happy' for happy-2.0.2..
[1 of 2] Compiling Paths_happy      ( dist/build/happy/autogen/Paths_happy.hs, dist/build/happy/happy-tmp/Paths_happy.o )
[2 of 2] Compiling Main             ( app/Main.lhs, dist/build/happy/happy-tmp/Main.o )
/home/runner/.cabal/lib/x86_64-linux-ghc-9.8.2/happy-lib-2.0.2-EgcIv144oLmqXufj3QhQk/Happy/Backend/LALR.hi
Declaration for lalrBackendDataDir
Unfolding of lalrBackendDataDir:
  getDataDir1 ErrorWithoutFlag
                Something is amiss; requested module  happy-lib-2.0.2-EgcIv144oLmqXufj3QhQk-backend-lalr:Paths_happy_lib differs from name found in the interface file happy-lib-2.0.2-EgcIv144oLmqXufj3QhQk-backend-glr:Paths_happy_lib (if these names look the same, try again with -dppr-debug)

<no location info>: error:
    Cannot continue after interface file error
```
So it looks like `v1-install` does not work for `happy` anymore.

Maybe we should reconsider using the interal libraries feature that cabal does not support fully yet.
- https://github.com/haskell/cabal/issues/8733

A proper support by cabal would be at least an error message stating that `v1-install` does not work with internal libraries.

**Comment by sgraf812** [comment]
I'm beginning to think that `happy-lib` should not use internal libraries. For an important ecosystem tool such as `happy`, there are just too many papercuts.

So I propose the following:

* Do not use internal libraries in `happy-lib`/vendor
* To guarantee separation of concerns, add another `happy-lib-sep.cabal` file which looks like the current `happy-lib.cabal` file. This file maintains the internal libraries separation and is built as part of CI. The net result is that we build everything twice, but since `happy-lib-sep.cabal` is not a dependency of the executable I hope that we don't need to build everything twice all that often (except when running `cabal test`).

**Comment by sgraf812** [comment]
A low maintenance alternative might be to instead add commented imports. Easier exemplified than explained.

* In `Happy.Tabular.Nameset`, add 

```hs
> import Happy.Grammar () -- See Note [Separation dependency]
```

(I see that `NameSet` already has this dependency. Bother.)

This dependency expresses that the "top" module in the dependency hierarchy of `happy-grammar` should be built before the "bottom" module of `happy-tabular`. (In general, there are no unique tops or bottoms, so perhaps this should rather be "sinks" and "sources".)

Another example: In `Happy.Backend.LALR`, add

```hs
> import Happy.Tabular () -- See Note [Separation dependency]
```

and so on.

Now, if there's ever a violation of the separation that the package declarations would enforce, we will get a cyclic import error.

**Comment by int-index** [comment]
v2- commands have been the default in cabal-install for a while now. I'm not at all worried about compatibility with v1- commands, which will eventually get removed/deprecated anyway.

**Comment by Ericson2314** [comment]
Yes in general I am sympathetic that we can only get away with saying "actually it's a Cabal bug" so many times, but in this case I agree with @int-index, I have no sympathy for anyone that insists on using v1 commands.

https://github.com/agda/agda/issues/4627#issuecomment-2377235920 I left this comment because I think the Agda comment is just plain out of date at this point.

**Comment by sgraf812** [comment]
I'm hesitant if it's a good idea for us to make an unforced change that in turn forces hundreds of Agda users to change their workflows... I would rather hear from other Agda users that they are fine with `v2-install`.

**Comment by Ericson2314** [comment]
@sgraf812 Well they already changed their CI? It's not clear to me at least that they want anything more from us.

**Comment by sgraf812** [comment]
Well, the comment you cited indicated that the Agda team recommends its users to install Agda with `v1-install`. Changing that recommendation would break a lot of habits and existing documentation.

Although upon searching in https://agda.readthedocs.io/en/latest/, there aren't that many mentions of `v1-install`.

As I said, we should ask people who know how many people are affected by this.

Ah, it may be that Agda users aren't really affected by this because `cabal v1-install Agda` will just use the `happy` in `$PATH`. Anyway, I'm confused enough to wait for a response from an Agdress (sorry, couldn't resist).

**Comment by Ericson2314** [comment]
Yeah I am fine waiting too. But my guess is this is basically a stale comment (that, according to the one person, linked the wrong issue too). I imagine we'll all be a bit confused at first :).

**Comment by andreasabel** [comment]
So yes, what I was trying to say is that it should not be a big problem for Agda users, as the official installation instructions do not recommend `v1-install` unless you want to install the development version.  But even then, it says to `cabal install happy` first which would not cause the present issue.
So we could hope that no more complaints come in...

**Comment by sgraf812** [comment]
I'm inclined to close this issue as won't fix then. Feel free to reopen if you think your `v1-install` use case is urgent.

**Comment by andreasabel** [comment]
The current behavior of `v1-install` is now (happy-2.1.4):
- cabal-3.12.1.0: succeeds
- cabal-3.14.1.1: fails
```
$ cabal v1-install happy 
Resolving dependencies...
Error: [Cabal-7654]
No cabal file found.
Please create a package description file <pkgname>.cabal

Failed to install happy-lib-2.1.4
Error: [Cabal-7140]
Some packages failed to install:
happy-2.1.4-FEt88s8bYtb4jlboxNtUeZ depends on happy-2.1.4 which failed to install.
happy-lib-2.1.4-1UOKRLhw4Nh6fWsmrPkI7e failed during the configure step. The exception was:
  ExitFailure 1
```

**Comment by andreasabel** [comment]
happy-2.1.4 closed this ticket.

> Given that this ticket got closed, is there also a fix for [haskell/cabal#10761](https://github.com/haskell/cabal/issues/10761)?

That one I reported with happy-2.1.4, but somehow cannot reproduce it, so I closed it now.

**Comment by Mikolaj** [comment]
Thank you.
